custom/icds_reports/utils/aggregation_helpers/distributed/agg_dashboard_activity.py

```diff
@@ -0,0 +1,275 @@
+                transformed_locations[district_id] = {
+                    'loc_level': 2,
+                    'parents': {
+                        'state_id': state_id
+                    }
+                }
+
+            if block_id not in transformed_locations:
+                transformed_locations[block_id] = {
+                    'loc_level': 3,
+                    'parents': {
+                        'district_id': district_id,
+                        'state_id': state_id
+                    }
+                }
+        return transformed_locations
+
+    def get_user_locations(self):
+        user_locations = list()
+        for user in self.dashboard_users:
+
+            state_id, district_id, block_id, user_level = None, None, None, None
+
+            if user['location_id']:
+                user_level = self.transformed_locations.get(user['location_id'])['loc_level']
+
+                if user_level == 1:
+                    state_id = user['location_id']
+                    district_id = 'All'
+                    block_id = 'All'
+                elif user_level == 2:
+                    state_id = self.transformed_locations.get(user['location_id'])['parents']['state_id']
+                    district_id = user['location_id']
+                    block_id = 'All'
+                elif user_level == 3:
+                    state_id = self.transformed_locations.get(user['location_id'])['parents']['state_id']
+                    district_id = self.transformed_locations.get(user['location_id'])['parents']['district_id']
+                    block_id = user['location_id']
+
+            user_locations.append((
+                user['username'],
+                state_id,
+                district_id,
+                block_id,
+                user_level
+            ))
+
+        return user_locations
+
+    @property
+    def tablename(self):
+        return "{}_{}".format(self.aggregate_parent_table, self.date.strftime("%Y-%m-%d"))
+
+    def drop_table_query(self):
+        return 'DROP TABLE IF EXISTS "{}"'.format(self.tablename)
+
+    def create_table_query(self):
+        return """
+        CREATE TABLE IF NOT EXISTS "{tablename}" (
+            CHECK (date = DATE %(date)s),
+            LIKE "{parent_tablename}" INCLUDING DEFAULTS INCLUDING CONSTRAINTS INCLUDING INDEXES
+        ) INHERITS ("{parent_tablename}")
+        """.format(
+            parent_tablename=self.aggregate_parent_table,
+            tablename=self.tablename,
+        ), {
+            "date": self.date.strftime("%Y-%m-%d"),
+        }
+
+    def get_last_agg_date(self):
+        from custom.icds_reports.models.aggregate import DashboardUserActivityReport
+        result = (DashboardUserActivityReport.objects.
+                  filter(date__lt=self.date.strftime("%Y-%m-%d")).
+                  aggregate(Max('date')))
+        if result:
+            return result['date__max']
+        return None
```

Comment:
this should probably return `date(1970,1,1)` or something instead of `None`. In fact won't this cause a query to fail if there are no reports in the database yet?

Reply:
No this works. If there is no report yet.